### PR TITLE
fix: append an empty expired message command

### LINKED ISSUE #10643: Message column families are not cleaned up and can cause outages
**Describe the bug**

I have now proof that the message column families in our mixed benchmarks are not cleaned up 
messages

These are the cause of https://github.com/camunda/zeebe/issues/10552#issuecomment-1272235756.

There is no way to clean up these column families, which is why I mark this bug as severity/critical. The state will grow at some point to a size, which will be no longer handlable by raft either.

**To Reproduce**

Run a mixed benchmark. You can run a benchmark with this branch [to get better insights](https://github.com/camunda/zeebe/commits/zell-investigate-mixed).

Especially https://github.com/camunda/zeebe/commit/547bf9c845b4b26ea3d4ceb375131f3d5b5c8c21 is interesting for the column family key count and https://github.com/camunda/zeebe/commit/8c134305f860703891010ad72c9c6332f950e0e2 for rocksdb statistics.

**Root Cause**

When appending a `expire` message command, the entire message including its variables is added to the appended command. And the commands are batched in a (default `maxMessageSize` of) 4MB buffer. 

https://github.com/camunda/zeebe/blob/720656b69cd28f5b4d14d516cad6f52a250c4107/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageTimeToLiveChecker.java#L37-L54

Depending on the size of the variables, a "big" chunk of memory will be already occupied for a single expire message command inside the buffer. Once the buffer is reached or exceeded, the `MessageTimeToLiveChecker` will stop collecting expired messages. 

In the mixed benchmark 100 messages per second are published to Zeebe. Each message is published with variables having a size of ~13kb (see `big_variables.json` - the actual json file size is ~46kb). Since the `MessageTimeToLiveChecker` runs every 60s, it must collect ~6k expired messages every 60s. With a partition size of 3, that means it must expire ~2k messages per partition. But with a variable size of ~13kb per message, the `MessageTimeToLiveChecker` can only collect ~300 expired messages every 60s. That's the point where Zeebe starts to build up state in the message column families. Consequently, the state grows infinitely, resulting in a significant performance drop.

**Expected behavior**

Column families are cleaned up correctly. Meaning, the expired messages are discarded at the same pace as they are published to Zeebe. To break down:
* Only the relevant message key is written to the expire message command (to save memory). That way, the checker can append up to 30k expire message commands in a 4MB buffer.
* Since only the message key is required, Zeebe can prevent doing an additional lookup to retrieve the actual stored message.

**Hints**
* Exporters:
  * By excluding irrelevant data in the command, the exported command to expire a message will only contain the message key.
  * That means, the correlation key, the variables, etc. are not part of any exported expire message command.
  * The same applies to the expired event that gets append to the log stream when the command got processed successfully.

**Environment:**
- OS: 
- Zeebe Version: 8.1.0, 8.2, (haven't checked for 8.0)
- Configuration:

### PR DESCRIPTION
**Pull request by romansmirnov**
## Description

To discard an expired message the processor and applier only need the message key. All other information is not necessary to process a expire message command. That's why only the `messageKey` is submitted to the command to expire a message.

This prevents the message checker to stop too early when collecting expired messages and not being able to collect as many as necessary expired messages in one batch to prevent building up state.

## Related issues
closes #10643 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/message/PublishMessageTest.java [hunk 1]
```diff
@@ -190,9 +190,9 @@ public void shouldExpireMessageAfterTTL() {
             .getFirst();
 
     Assertions.assertThat(deletedEvent.getValue())
-        .hasName("order canceled")
-        .hasCorrelationKey("order-123")
-        .hasTimeToLive(100L)
+        .hasName("")
+        .hasCorrelationKey("")
+        .hasTimeToLive(-1L)
         .hasMessageId("");
```
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/message/PublishMessageTest.java:196]: ❌ This test doesn't verify that we delete the correct message, as we don't assert the key. We should add an assertion:

`Assertions.assertThat(deletedEvent).hasKey(publishedRecord.getKey());`

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks for this detailed analysis @romansmirnov, very easy to understand!

I have one comment regarding a missing assertion. I also noticed that if the TTL is <= 0 we [still write the full message](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessagePublishProcessor.java#L123-L126). It doesn't have an impact on the issue you are fixing with this PR, so I'm not worried about it. Just wanted to point this out.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/message/PublishMessageTest.java (hunk 1 above)
Comment (romansmirnov) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/message/PublishMessageTest.java]: @remcowesterhoud, thanks for raising it. I adjusted the test case accordingly 69ef1e9da1cbca03fcbd70419699cac793f24610

> I also noticed that if the TTL is <= 0 we [still write the full message](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessagePublishProcessor.java#L123-L126).

**Review by remcowesterhoud — APPROVED** [review]
Thanks for the change @romansmirnov, LGTM!

> Great catch. I haven't seen it. Let me know if you want me to change that in a separate PR.

I'm fine with keeping it as-is. It doesn't have an impact on the bug fix and it's easier to just pass the full record since we have it available there.

bors merge

**Comment by romansmirnov** [comment]
bors merge

**Comment by remcowesterhoud** [comment]
bors retry

**Comment by remcowesterhoud** [comment]
bors cancel

**Comment by remcowesterhoud** [comment]
bors merge

**Pull request merged**